### PR TITLE
Improve access control in peer connections

### PR DESCRIPTION
**Pull request by yyforyongyu**
We now remove the restriction placed on outbound connections - previously we would account them for `numRestricted`, which is no longer the case, as for user-initiated connections, it should not be put under the same restriction as the inbound ones - we may choose to eventually put a restriction on the outbound too, but it should be a very different category from the inbound.

In addition more tests are patched to fix the case that an existing connection may be counted multiple times, causing the `numRestricted` to be falsely reached.

Depends on,
- #9876 

TODOs:
- [x] Add more tests
- [x] add detailed descriptions

- - -
This change is [Reviewable](https://reviewable.io/reviews/lightningnetwork/lnd/9880)

### CODE REVIEW

File: accessman.go [hunk 1]
```diff
@@ -138,6 +168,12 @@ func (a *accessMan) assignPeerPerms(remotePub *btcec.PublicKey) (
 	// peer.
 	acsmLog.DebugS(ctx, "Peer has no channels, assigning restricted access")
 
+	// If this is an existing peer, there's no need to check for slot limit.
```
Comment (ziggie1984) [inline accessman.go:171]: why do we call this after the disconnect  check, I think it makes sense to do it before ?

**Comment by litbot-9000** [comment]
@ziggie1984: review reminder

File: server.go [hunk 2]
```diff
@@ -4980,7 +4980,11 @@ func (s *server) ConnectToPeer(addr *lnwire.NetAddress,
 
 	// Ensure we're not already connected to this peer.
 	peer, err := s.findPeerByPubStr(targetPub)
-	if err == nil {
+
+	// When there's no error it means we already have a connection with this
+	// peer. If this is a dev environment with the `--unsafeconnect` flag
+	// set, we will ignore the existing connection and continue.
+	if err == nil && !s.cfg.Dev.GetUnsafeConnect() {
```
Comment (ziggie1984) [inline server.go:5006]: So what happens now that we allow 2 connections simultaneously? Is the newer one more prioritised ?

File: itest/lnd_access_perm_test.go [hunk 3]
```diff
@@ -1,93 +1,341 @@
...
+	{
+		Name:     "restricted on inbound",
+		TestFunc: testPeerConnRestrictedOnInbound,
+	},
+	{
+		Name:     "already connected",
+		TestFunc: testPeerConnAlreadyConnected,
+	},
+	{
+		Name:     "unlimited outbound",
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:18]: we have this itest here, but in the previous commits I have not seen that we removed the restriction for now ?

File: itest/lnd_access_perm_test.go [hunk 4]
```diff
@@ -1,93 +1,341 @@
...
+	ht.AssertConnected(alice, carol)
+	ht.AssertNotConnected(alice, bob)
+}
+
+// testPeerConnAlreadyConnected checks that when there's already a connection
+// alive, another attempt to connect doesn't take up the slot or kill the
+// existing connection, in specific,
+//   - When Alice has an inbound connection from Bob, another inbound connection
+//     from Bob will be noop.
+//   - When Bob has an outbound connection to Alice, another outbound connection
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:79]: What about the Case Bob has an outbound and Alice tries to connect via a new Inbound ?

File: accessman.go [hunk 5]
```diff
@@ -483,6 +483,8 @@ func (a *accessMan) newOpenChan(remotePub *btcec.PublicKey) error {
 // encoded key, we should not accept this incoming connection or immediately
 // disconnect. This does not assign to the server's peerScores maps. This is
 // just an inbound filter that the brontide listeners use.
+//
+// TODO(yy): this check is redundant, remove it.
```
Comment (ziggie1984) [inline accessman.go]: Nit: Could you explain why this is redundant here and reference the part which does does the same thing making this one essentially obsolete ?

**Comment by ziggie1984** [comment]
@Crypt-iQ could you take a look at this new design where we remove outbound restrictions, are there some downsides with the current LND codebase which we do not see, referring to your comment here:

https://github.com/lightningnetwork/lnd/issues/9874#issuecomment-2968514790

File: server.go (hunk 2 above)
Comment (ziggie1984) [reply server.go:5006]: Still not quite sure why we need this here, will the peer simply be rejected on layer downwards in the `connmgr` ?

File: itest/lnd_access_perm_test.go (hunk 3 above)
Comment (ziggie1984) [reply itest/lnd_access_perm_test.go:18]: This commit should land after we relax the outbound connections.

File: itest/lnd_access_perm_test.go [hunk 6]
```diff
@@ -1,93 +1,341 @@
...
+	{
+		Name:     "already connected",
+		TestFunc: testPeerConnAlreadyConnected,
+	},
+	{
+		Name:     "unlimited outbound",
+		TestFunc: testPeerConnUnlimitedOutbound,
+	},
+	{
+		Name:     "upgrade perm",
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go]: There are now 2 variables in the code base which are named `perm`, one stands for permanent the other for permission, I would propose using more clear variable names.

File: itest/lnd_access_perm_test.go [hunk 7]
```diff
@@ -1,93 +1,341 @@
...
+// existing connection, in specific,
+//   - When Alice has an inbound connection from Bob, another inbound connection
+//     from Bob will be noop.
+//   - When Bob has an outbound connection to Alice, another outbound connection
+//     to Alice will be noop.
+//
+// In this test we will create nodes using the dev flag `unsafeconnect` to mimic
+// the behaviour of persistent/permanent peer connections, where it's likely
+// inbound and outbound connections are happening at the same time.
+func testPeerConnAlreadyConnected(ht *lntest.HarnessTest) {
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:85]: I don't understand this test, you are saying we use `persistent/permanent peer` behaviour but I cannot see that there is a channel between those two peers or that you use the permanent flag when trying to connect to the peer.

File: itest/lnd_access_perm_test.go [hunk 8]
```diff
@@ -1,93 +1,341 @@
...
+	// Bob now connects to Alice - from Alice's PoV, this is her inbound
+	// connection from Bob. We expect Bob to connect to Alice successfully.
+	_, err := ht.ConnectNodesNoAssert(bob, alice)
+	require.NoError(ht, err)
 	ht.AssertConnected(alice, bob)
 
-	ht.ConnectNodes(alice, carol)
+	// Assert Alice's slot has been filled up by connecting Carol to Alice.
+	_, err = ht.ConnectNodesNoAssert(carol, alice)
+	require.NoError(ht, err)
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:140]: Looks a bit weird that we initiate the connection, require no error but are still not connected. On the other hand we should at least wait a bit so that Carol for sure will not be connected otherwise this quick check seems to make not real sense because before building up the connection Carol is not connected anyways.

File: itest/lnd_access_perm_test.go [hunk 9]
```diff
@@ -1,93 +1,341 @@
...
+
+	// Connect Carol to Alice - since Bob now has a pending channel with
+	// Alice, there's a slot available for Carol to connect.
+	ht.ConnectNodes(carol, alice)
+
+	// Open a channel from Carol to Alice and let it stay pending.
+	pendingUpdate = ht.OpenChannelAssertPending(carol, alice, p)
+	cpCA := lntest.ChanPointFromPendingUpdate(pendingUpdate)
+
+	// Mine the funding txns.
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:191]: if it is only the funding tx why do we expect 2 transactions here ?

File: itest/lnd_access_perm_test.go [hunk 10]
```diff
@@ -1,93 +1,341 @@
...
+	ht.MineBlocksAndAssertNumTxes(1, 2)
+
+	// Dave should be able to connect to Alice.
+	ht.ConnectNodes(dave, alice)
+
+	// Open a channel from Alice to Dave, which will free up Alice's slot.
+	cpAD := ht.OpenChannel(alice, dave, p)
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:198]: what does the `p` stand for in cpAD, OpenChannel already mines enough blocks so that it is not `pending` ?

File: itest/lnd_access_perm_test.go [hunk 11]
```diff
@@ -1,93 +1,341 @@
...
+
+// testPeerConnNodeRestart checks that when a peer has or used to have a channel
+// with Alice, when Alice restarts, she should still have the available slot for
+// more inbound connections.
+func testPeerConnNodeRestart(ht *lntest.HarnessTest) {
+	args := []string{"--num-restricted-slots=1"}
+
+	// Create a new node with one slot available.
+	alice := ht.NewNodeWithCoins("Alice", args)
+	bob := ht.NewNode("Bob", []string{"--dev.unsafeconnect"})
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:217]: why are we using this `unsafeconnect` only for bob ?

File: itest/lnd_access_perm_test.go [hunk 12]
```diff
@@ -1,93 +1,341 @@
...
+	// connection from Bob. We expect Bob to connect to Alice successfully.
+	_, err := ht.ConnectNodesNoAssert(bob, alice)
+	require.NoError(ht, err)
+	ht.AssertConnected(alice, bob)
+
+	// Carol now connects to Alice - from Alice's PoV, this is her inbound
+	// connection from Carol. Carol's connection should be failed.
+	_, err = ht.ConnectNodesNoAssert(carol, alice)
+	require.NoError(ht, err)
+	ht.AssertNotConnected(alice, carol)
```
Comment (ziggie1984) [inline itest/lnd_access_perm_test.go:60]: same as above, those nodes are not connected in the first place so testing that they are not after calling connected might be a race where we just not yet connected.

File: accessman_test.go [hunk 13]
```diff
@@ -250,9 +251,8 @@ func TestAssignPeerPerms(t *testing.T) {
 			expectedErr:      ErrGossiperBan,
 		},
 		// peer6 has no channel with us, and we expect it to have a
-		// restricted status. We also expect the error
-		// `ErrNoMoreRestrictedAccessSlots` to be returned given
-		// we only allow 1 restricted peer in this test.
+		// restricted status. Since this peer is seen, we don't expect
```
Comment (ziggie1984) [inline accessman_test.go:280]: Q: What do you mean by "the peer is seen", the peer is connected ?

File: accessman_test.go [hunk 14]
```diff
@@ -394,3 +394,135 @@ func TestAssignPeerPermsBypassRestriction(t *testing.T) {
...
+		require.NoError(t, err)
+
+		return peerPriv.PubKey()
+	}
+
+	// peer1 exists in `peerCounts` map.
+	peer1 := genPeerPub()
+	peer1Str := string(peer1.SerializeCompressed())
+
+	// peer2 exists in `peerScores` map.
```
Comment (ziggie1984) [inline accessman_test.go:441]: is it possible for a peer to exist in peerScore but not in peerCount ?

File: accessman_test.go [hunk 15]
```diff
@@ -394,3 +394,135 @@ func TestAssignPeerPermsBypassRestriction(t *testing.T) {
...
+	a, err := newAccessMan(cfg)
+	require.NoError(t, err)
+
+	// Add peer2 to the `peerScores`.
+	a.peerScores[peer2Str] = peerSlotStatus{
+		state: peerStatusTemporary,
+	}
+
+	// Assigning to peer1 should not return an error.
+	status, err := a.assignPeerPerms(peer1)
```
Comment (ziggie1984) [inline accessman_test.go:474]: you mention in the test that preexisting peer will not take a slot from the restricted count, but why is this peer restircted ?

File: accessman.go [hunk 16]
```diff
@@ -483,6 +483,8 @@ func (a *accessMan) newOpenChan(remotePub *btcec.PublicKey) error {
 // encoded key, we should not accept this incoming connection or immediately
```
Comment (ziggie1984) [inline accessman.go:485]: Nit: Personally I think we should not add some TODOs in the commit msg, we should state there what this commit is doing and add TODOs in the code part itself.

File: accessman.go [hunk 17]
```diff
@@ -497,36 +499,45 @@ func (a *accessMan) checkIncomingConnBanScore(remotePub *btcec.PublicKey) (
 	a.banScoreMtx.RLock()
```
Comment (ziggie1984) [inline accessman.go:503]: Q: maybe we rename the function to CheckPermission don't see that we compare banning scores ?

File: server.go [hunk 18]
```diff
@@ -1889,6 +1889,8 @@ func newServer(_ context.Context, cfg *Config, listenAddrs []net.Addr,
 		// connection requests when we call NewListener.
 		listeners[i], err = brontide.NewListener(
 			nodeKeyECDH, listenAddr.String(),
+			// TODO(yy): remove this check and unify the inbound
+			// connection check inside `InboundPeerConnected`.
 			s.peerAccessMan.checkIncomingConnBanScore,
```
Comment (ziggie1984) [inline server.go]: @Crypt-iQ what was our intention again to put this check earlier into the `HandShake` method rather than into the `InboundPeerConnected` ?

File: server.go [hunk 19]
```diff
@@ -4345,12 +4310,33 @@ func (s *server) notifyFundingTimeoutPeerEvent(op wire.OutPoint,
...
 
 	brontideConn := conn.(*brontide.Conn)
 	addr := conn.RemoteAddr()
 	pubKey := brontideConn.RemotePub()
 
+	// If the remote node's public key is banned, drop the connection.
+	//
+	// TODO(yy): Consider perform this check in
+	// `peerAccessMan.addPeerAccess`.
+	access, err := s.peerAccessMan.assignPeerPerms(pubKey)
```
Comment (ziggie1984) [inline server.go:4353]: I think we need to unify `addPeerAccess` and `assignPeerPerms ` because `peerConnected` can be called concurrently for many peers and we should increment the restricted Peers number if we are connecting the peer otherwise because we need to hold the`banMutex` it can be possible that we allow more restricted peers than possible.

2 peers call `peerConnected` both of these calls succeed with `assignPeerPerms` before one of them has called `addPeerAccess` to increase the counter of the restictive peer.

File: docs/release-notes/release-notes-0.19.2.md [hunk 20]
```diff
@@ -35,6 +35,10 @@
 
 ## Functional Updates
 
+- [Improved](https://github.com/lightningnetwork/lnd/pull/9880) the connection
```
Comment (ziggie1984) [inline docs/release-notes/release-notes-0.19.2.md:41]: Nit: Improve the connection ...

File: server.go (hunk 2 above)
Comment (yyforyongyu) [reply server.go:5006]: This is a dev flag only used in the itest, so not sure what it has to do with the race connection.

File: itest/lnd_access_perm_test.go (hunk 3 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go:18]: Moved - also just fyi this approach is called [TDD](https://en.wikipedia.org/wiki/Test-driven_development), which I will use a lot in the upcoming PRs.

File: itest/lnd_access_perm_test.go (hunk 6 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go]: updated - tho I don't think people would think of `upgrade permanent` when seeing `upgrade perm`

File: itest/lnd_access_perm_test.go (hunk 7 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go:85]: I think it made it clear in the comments - we use the dev flag to **mimic** the persistent/permanent peer behavior, so think it as a mock.

File: itest/lnd_access_perm_test.go (hunk 8 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go:140]: > Looks a bit weird that we initiate the connection, require no error but are still not connected.

That is RPC layer error - ofc we can change the RPC to return an error here, but it means blocking. Are you sure we wanna do that?

File: itest/lnd_access_perm_test.go (hunk 12 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go:60]: will see what i can do - maybe a sleep somewhere

File: itest/lnd_access_perm_test.go (hunk 9 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go:191]: there are two

File: itest/lnd_access_perm_test.go (hunk 10 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go:198]: **c**hannel**p**oint - i think it's used everywhere in the itest.

File: itest/lnd_access_perm_test.go (hunk 11 above)
Comment (yyforyongyu) [reply itest/lnd_access_perm_test.go:217]: because it's bob that tried to connect to alice multiple times.

File: accessman.go (hunk 1 above)
Comment (yyforyongyu) [reply accessman.go:171]: if the peer is messing with us we should disconnect?

File: accessman_test.go (hunk 13 above)
Comment (yyforyongyu) [reply accessman_test.go:280]: that the peer is found in one of the two maps.

File: accessman_test.go (hunk 14 above)
Comment (yyforyongyu) [reply accessman_test.go:441]: I think so - at lease for a short period of time, the peer gets into `peerScore` first, then `peerCount`, that's the lifecycle of the peer.

File: accessman_test.go (hunk 15 above)
Comment (yyforyongyu) [reply accessman_test.go:474]: no `ErrNoMoreRestrictedAccessSlots` is returned here, so it's not taking the slot - note that the peer's access status and whether it takes a slot or not are independent here.

File: accessman.go (hunk 16 above)
Comment (yyforyongyu) [reply accessman.go:485]: updated

File: accessman.go (hunk 5 above)
Comment (yyforyongyu) [reply accessman.go]: updated

File: accessman.go (hunk 17 above)
Comment (yyforyongyu) [reply accessman.go:503]: cool updated

File: server.go (hunk 19 above)
Comment (yyforyongyu) [reply server.go:4353]: I think whenever we review PRs it's good to keep in mind of two things - first, what is the scope of the PR, which the author usually states it in the PR description. Second, what is the issue identified - is it preexisting, or it's introduced by this PR. If a new PR introduced a buggy behavior, that def needs to be addressed. If it's preexisting, then there's no need to fix everything in one PR, as long as the original scope is achieved. This is to make sure we move forward with progress, and not putting all the task under one author.

As for this issue, I think it's preexisting, otherwise I won't bother creating that TODO. In fact, since we are now moving the `assignPeerPerms` here instead of outside of `peerConnected`, it already makes the situation better.

**Comment by yyforyongyu** [comment]
> I think we also need to think about the two functions:

Cool think this is now fixed.

File: itest/lnd_access_perm_test.go (hunk 3 above)
Comment (ziggie1984) [reply itest/lnd_access_perm_test.go:18]: Makes sense, I think what would be cool to just state it in that commit that we introduce some tests which won't pass. But seems like a cool strategy TDD.

File: accessman.go (hunk 1 above)
Comment (ziggie1984) [reply accessman.go:171]: Yeah I was more looking into the shouldDisconnect logic and saw there that we are already do not disconnect if we have a channel so yeah maybe it is better for the accessMan not assume stuff happening in other subsystems:

https://github.com/lightningnetwork/lnd/blob/31c74f20fa46259dcd351ba277a9a216fc6edcdc/discovery/gossiper.go#L3710

Just wanted to bring it up.

File: accessman_test.go (hunk 15 above)
Comment (ziggie1984) [reply accessman_test.go:474]: Ok now I understand it, but shouldn't we make sure we remove peers from the peerCount as well which have no channels in case we remove the peer in `removePeerAccess`.

Otherwise all peers we ever connected outbound wise will have access when connecting inbound ?

File: server.go (hunk 19 above)
Comment (ziggie1984) [reply server.go:4353]: Agree we should not fix it in this PR. But I think reviews like this are still a good forum to discuss potential problems of the current code. I was not aware of this race condition beforehand that's why I brought it up and wanted to get your opinion here as well.

File: itest/lnd_access_perm_test.go (hunk 8 above)
Comment (ziggie1984) [reply itest/lnd_access_perm_test.go:140]: Just wanted to bring it to your attention, let's keep it as is.

File: itest/lnd_access_perm_test.go (hunk 7 above)
Comment (ziggie1984) [reply itest/lnd_access_perm_test.go:85]: Ok now I understand, nice edge case test, didn't understand it at first.

File: funding/manager.go [hunk 21]
```diff
@@ -511,7 +511,7 @@ type Config struct {
 
 	// NotifyOpenChannelEvent informs the ChannelNotifier when channels
 	// transition from pending open to open.
-	NotifyOpenChannelEvent func(wire.OutPoint, *btcec.PublicKey) error
+	NotifyOpenChannelEvent func(wire.OutPoint, *btcec.PublicKey)
```
Comment (ziggie1984) [inline funding/manager.go:514]: Thank you for fixing this 🙏

File: accessman.go [hunk 22]
```diff
@@ -77,6 +81,48 @@ func newAccessMan(cfg *accessManConfig) (*accessMan, error) {
...
+
+			// Exit early if the peer is no longer restricted.
+			return peerStatusTemporary, true
+		}
+
+		return peerStatusRestricted, true
+	}
+
+	// Check if the peer is found in the scores map.
+	status, found := a.peerScores[pub]
```
Comment (Roasbeef) [inline accessman.go:115]: Appears to also be a bug fix as we didn't check to see if has an existing score entry before?

File: accessman_test.go (hunk 15 above)
Comment (Roasbeef) [reply accessman_test.go:474]: IIUC, we remove a peer's score when they disconnect: https://github.com/lightningnetwork/lnd/blob/168d63ce376947c2dbb0de880e3cdae8ed047e24/server.go#L4960-L4961

So if we connect to a peer that we have no channels with, then we disconnect and they reconnect inbound, they'll start with a blank score.

File: accessman.go [hunk 23]
```diff
@@ -497,36 +501,45 @@ func (a *accessMan) checkIncomingConnBanScore(remotePub *btcec.PublicKey) (
...
+	if found {
+		acsmLog.DebugS(ctx, "Peer found (protected/temporary), "+
+			"accepting")
 
-			return true, nil
-		}
+		return true, nil
+	}
+
+	_, found = a.peerScores[peerMapKey]
```
Comment (Roasbeef) [inline accessman.go:516]: Hmm, perhaps we should consider unifying `peerScores` and `peerCounts`, or creating some methods that do composite checks against both maps.

Comment (Roasbeef) [reply accessman.go:516]: Either way, I think we should consider renaming `peerCounts` to `peerHistoticalChannelCounts` or something like that.

File: accessman.go [hunk 24]
```diff
@@ -440,6 +440,8 @@ func (a *accessMan) newOpenChan(remotePub *btcec.PublicKey) error {
 		}
 
 		peerCount.HasOpenOrClosedChan = true
+		peerCount.PendingOpenCount -= 1
```
Comment (Roasbeef) [inline accessman.go:443]: Can we update the tests to cover this case?

File: accessman.go (hunk 22 above)
Comment (yyforyongyu) [reply accessman.go:115]: yeah exactly

File: accessman.go (hunk 1 above)
Comment (yyforyongyu) [reply accessman.go:171]: > it is better for the accessMan not assume stuff happening in other subsystems

yeah good thought - that's exactly the decoupling we are aiming for, which means we will need a design doc for this, to explicitly picture how these subsystems interact with each other.

File: server.go (hunk 19 above)
Comment (yyforyongyu) [reply server.go:4353]: Yeah def - we need to bring awareness so it's good to discuss the potential issue. Tho based on offline discussion it seems to me that this is a blocking comment, which is why I brought it up.

File: funding/manager.go (hunk 21 above)
Comment (yyforyongyu) [reply funding/manager.go:514]: Thank you for spotting this!

File: accessman.go (hunk 23 above)
Comment (yyforyongyu) [reply accessman.go:516]: Done - renamed it in the last commit.

File: accessman.go (hunk 24 above)
Comment (yyforyongyu) [reply accessman.go:443]: Updated and patched the test.

**Comment by yyforyongyu** [comment]
> Should we remove the Peer from the PeerCount map in case it has not open/pending channel. Otherwise Peers will always be able to reconnect as inbound although we disconnected to them ?

Yeah good catch - added a commit plus some tests to remove it, but only if the peer has no channels with us.

File: accessman_test.go [hunk 25]
```diff
@@ -703,3 +703,111 @@ func TestAddPeerAccessOutbound(t *testing.T) {
...
+	// Remove peer1 should change nothing.
+	a.removePeerAccess(peer1)
+
+	// peer1 should be removed from peerScores but not peerChanInfo.
+	_, found := a.peerScores[peer1]
+	require.False(t, found)
+	_, found = a.peerChanInfo[peer1]
+	require.True(t, found)
+
+	// Remove peer1 should change nothing.
```
Comment (ziggie1984) [inline accessman_test.go]: Nit: peer2

File: accessman_test.go [hunk 26]
```diff
@@ -703,3 +703,111 @@ func TestAddPeerAccessOutbound(t *testing.T) {
...
+
+	// Create a testing accessMan.
+	a := &accessMan{
+		peerChanInfo: make(map[string]channeldb.ChanCount),
+		peerScores:   make(map[string]peerSlotStatus),
+	}
+
+	// numRestrictedExpected specifies the final value to expect once the
+	// test finishes.
+	var numRestrictedExpected int
```
Comment (ziggie1984) [inline accessman_test.go:720]: Nit: Maybe start from 1 here so we do not decrement to -1 ?

**Review by ziggie1984 — APPROVED** [review]
LGTM 🥽



sample.conf would need an update now as well which currently states:
```
; The number of restricted slots the server will allocate for peers.
; num-restricted-slots=100
```

Something along the lines like:

This restriction only counts for inbound however outbound are still counted as restricted, so you can have more than 100 restricted peers, but not more inbound peers if the number exceeded the max-restricted.

File: accessman_test.go (hunk 26 above)
Comment (yyforyongyu) [reply accessman_test.go:720]: yeah i kinda wanna show that this num can be negative as we need underflow protection.

**Comment by yyforyongyu** [comment]
> sample.conf would need an update now as well which currently states:

@ziggie1984 Updated - note that outbound connections are not restricted, so they should not increment the slots count `numRestricted` since they would have `peerStatusTemporary` by default.

**Review by Roasbeef — APPROVED** [review]
:lgtm:

> Reviewed 2 of 11 files at r1, 9 of 9 files at r2, 8 of 11 files at r5, 5 of 5 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/lightningnetwork/lnd/9880#-:-OTER_A81e2e-SnLUeJN:bnfp4nl)* status: all files reviewed, 26 unresolved discussions (waiting on @Crypt-iQ, @yyforyongyu, and @ziggie1984)

**Comment by Roasbeef** [comment]
Accidentally dismissed the existing approval, merging bsaed on that.
